### ISSUE TITLE
ClassGraphFacade fails for inner classes

### ISSUE THREAD

**Issue by peter721** [j-easy/easy-random#353]
Class.getCanonicalName() and Class.getName() generally return the same thing - except for inner classes. ClassGraph will find inner classes based on their "name", so the code in "searchForPublicConcreteSubTypesOf" will fail when string-matching the canonicalName.

**Comment by benas** [comment]
Do you have a failing test?

If I understand correctly, you mean we need to use `type.getName()` instead of `type.getCanonicalName()` here: https://github.com/j-easy/easy-random/blob/master/easy-random-core/src/main/java/org/jeasy/random/util/ClassGraphFacade.java#L56

Is that correct?

**Comment by benas** [comment]
Hi @peter721 ,

> so the code in "searchForPublicConcreteSubTypesOf" will fail when string-matching the canonicalName.

I'm not able to reproduce the issue (unless I misunderstand it). Here is a **passing** test with version `4.0.0.RC1`:

```java
package org.jeasy.random;

public interface IBar {
}
```

```java
package org.jeasy.random;

import org.assertj.core.api.Assertions;
import org.junit.jupiter.api.Test;

public class Foo {

    private IBar bar;

    public IBar getBar() {
        return bar;
    }

    public class Bar implements IBar {

    }

    @Test
    void testScanClasspathForConcreteTypes_whenSubTypeIsAnInnerClass() {
        EasyRandomParameters parameters = new EasyRandomParameters()
                .scanClasspathForConcreteTypes(true);
        EasyRandom easyRandom = new EasyRandom(parameters);

        Foo foo = easyRandom.nextObject(Foo.class);

        Assertions.assertThat(foo).isNotNull();
        Assertions.assertThat(foo.getBar()).isInstanceOf(Bar.class);
    }
}
```

In this example, `Foo.Bar.class.getName() = org.jeasy.random.Foo$Bar` and `Foo.Bar.class.getCanonicalName() = org.jeasy.random.Foo.Bar`, and still ClassGraph is able to find the concrete type `Bar` **by its canonical name** even if it is an inner class.

Do you agree?

**Comment by peter721** [comment]
Hi, thanks for the reply!

The ClassGraphFacade.searchForPublicConcreteSubTypesOf() method must be called with an inner type to illustrate the problem. The following code does fail:
```java
public class Foo {

  public abstract class Bar {}

  public class Bar2 extends Bar {}

  public class Baz {
    private Bar bar;

    public Bar getBar() {
      return bar;
    }
  }

  @Test
  void testScanClasspathForConcreteTypes_whenConcreteTypeIsAnInnerClass() {
    EasyRandomParameters parameters =
        new EasyRandomParameters().scanClasspathForConcreteTypes(true);
    EasyRandom easyRandom = new EasyRandom(parameters);

    Foo.Baz baz = easyRandom.nextObject(Foo.Baz.class);

    Assertions.assertThat(baz).isNotNull();
    Assertions.assertThat(baz.getBar()).isNotNull();
  }
}
```

**Comment by benas** [comment]
Thank you for the clarification. The issue happens when **both** the abstract type and the concrete one are inner classes. I fixed the issue by using the type name instead of its canonical name.

The fix will be part of the upcoming 4.0.0 release. Stay tuned.

Thank you for reporting this issue just in time 😄
